WIP: Tom/build g4 v11

TomGoffrey opened the pull request:
Opening for discussion.

TomGoffrey committed:
Roll back to G4 10.7 for initial build

TomGoffrey committed:
Register step limiter

TomGoffrey committed:
Initial copy of files from Boris integration example

TomGoffrey committed:
Update CMakeLists.txt

TomGoffrey committed:
Remove tests from CMake [temp]

TomGoffrey committed:
Require Geant4 gdml support via CMake

TomGoffrey committed:
Drop 10.X support

TomGoffrey committed:
Fix typos

TomGoffrey committed:
Add missing file

TomGoffrey committed:
Fix include

TomGoffrey committed:
Missing declarations

TomGoffrey committed:
Fix typos

TomGoffrey committed:
Add missing include statement

TomGoffrey committed:
Fix some inherited typos from miniapp

TomGoffrey committed:
Fixes to run action

TomGoffrey commented on src/QTRunAction.cc:
```diff
@@ -53,15 +52,15 @@ EGRunAction::EGRunAction(QTEventAction* eventAction, G4String name, G4int na)
   for (G4int i=0;i<nAntenna;++i) {
-    analysisManager->CreateNtupleTColumn(tvecname + std::to_string(i), fEventAction->GetTimeVec(i));
-    analysisManager->CreateNtupleTColumn(vvecname + std::to_string(i), fEventAction->GetVoltageVec(i));
+    analysisManager->CreateNtupleDColumn(tvecname + std::to_string(i));
```
@yramachers I'm not convinced my "fixes" here are correct. It looks like you're trying to record a name-value pair here? Is that intentional - `CreateNtupleTColumn` doesn't appear to be a member of any of the Analysis Manager classes, and it's not used in any of the examples. I've dug a bit into the source code of `G4NtupleBookingManager.cc` and it doesn't appear to match what you're doing here either?

yramachers replied:
@TomGoffrey , I agree; it looks like the TColumn as a template column creation is a private method of the booking manager. Don't remember why I used it explicitly other than making sure that we need to store a full std::vector here in the column. It is possible to use the storage of values in the vector but that would be another loop over vector content. This ability to store a vector as single entry in a row was enabled in older versions, then disappeared, then apparently shows up again, confusing. 
Nevertheless, storing a full vector as a single row entry is extremely useful for the output file.
So, yes, DColumn will do but I think we need to use the booking manager method

G4int CreateNtupleDColumn(const G4String& name, std::vector<double>* vector);

which ought to set up storage of a std::vector.

TomGoffrey replied:
Thanks!

TomGoffrey replied:
Adding discussion at:
https://geant4-forum.web.cern.ch/t/filling-std-vector-type-ntuples-using-g4root/2118

for reference. See also recent FIXME comments.

TomGoffrey commented on src/QTRunAction.cc:
```diff
@@ -18,7 +18,6 @@ EGRunAction::EGRunAction(QTEventAction* eventAction, G4String name, G4int na)
   auto analysisManager = G4AnalysisManager::Instance();
 
   // Create directories
-  analysisManager->SetDefaultFileType("root");
```
I believe this is no longer necessary in v11.X - file types are inferred from extensions. Tagging @TomGoffrey @yramachers

yramachers replied:
Ok, that was copy/paste from older versions. Would be good if that can disappear.

TomGoffrey committed:
Remove argument from destructor

TomGoffrey committed:
Fix typo in include statement

TomGoffrey commented on src/QTTrajectory.cc:
```diff
@@ -74,11 +74,11 @@ std::pair<double,double> QTTrajectory::convertToVT(unsigned int which)
-  G4double omega = pfieldmanager->GetChordFinder()->GetIntegrationDriver()->GetEquationOfMotion()->GetOmega().mag();
+  G4ThreeVector Bfield = pfieldmanager->GetChordFinder()->GetIntegrationDriver()->GetEquationOfMotion()->GetCachedFieldValue();
+  G4double omega = pfieldmanager->GetChordFinder()->GetIntegrationDriver()->GetEquationOfMotion()->CalcOmegaGivenB(Bfield).mag();
+  acc = pfieldmanager->GetChordFinder()->GetIntegrationDriver()->GetEquationOfMotion()->CalcAccGivenB(Bfield);
```
@DSwinnockPhysics Could I just check exactly what acceleration should be returned here? Is it just the Lorentz force, or Lorentz+radiation reaction?

DSwinnockPhysics replied:
Based on the code I have (which uses Seb's code), it's using Lorentz+radiation reaction.

DSwinnockPhysics replied:
Also I noticed later on in this code (QTTrajectory) I think there is a factor of c missing in the near field calculation since it's being divided by c in the fac variable and in the equation from Duke only the far field is divided by c.

TomGoffrey committed:
Make QTTrajectory a concrete class

TomGoffrey committed:
Fix constructor declaration

TomGoffrey committed:
Add missing include statements

TomGoffrey committed:
Add TBetaGenerator.hpp

TomGoffrey commented on include/TBetaGenerator.hpp:
```diff
@@ -0,0 +1,363 @@
+// Collect all physical and utility constants in one specific place
```
@yramachers As ever, I'm torn between directly including the code (ease of use for users, no forgetting of `--recursive` flag) vs including as a submodule. For now I'm going with direct inclusion, but I'm happy to discuss/entertain possible alternative viewpoints.

TomGoffrey committed:
11.X analysis changes

TomGoffrey committed:
11.X analysis changes

TomGoffrey committed:
Fix typo

TomGoffrey committed:
Fix typo

TomGoffrey committed:
Fix typo

TomGoffrey committed:
Fix typo

TomGoffrey committed:
Fix typos in Event Action

TomGoffrey committed:
Fix typos in Event Action

TomGoffrey committed:
Fix/Add FIXMEs for vector output

TomGoffrey committed:
Fix/Add FIXMEs for vector output

TomGoffrey committed:
Set-up vector output in QTRunAction

TomGoffrey committed:
Fixes to QTPrimaryGeneratorAction

TomGoffrey committed:
Compilation fixes to trajectory

TomGoffrey committed:
Compilation fixes for TrackingAction

TomGoffrey committed:
Placeholder code to allow compilation

TomGoffrey committed:
Don't use cached field values

(For Boris these are unlikely to be at the correct position due to
leapfrog method)

TomGoffrey committed:
Add omega calculation

TomGoffrey committed:
Add acceleration calculation

TomGoffrey committed:
Standards compliant pi

TomGoffrey committed:
Fix typo

TomGoffrey committed:
Make Gamma inline

TomGoffrey committed:
Delete old no longer valid stepper types

TomGoffrey committed:
Always use custom equation of motion in field set-up

TomGoffrey committed:
Fix inheritance of QTEquationOfMotion

yramachers committed:
Merge branch 'tom/build_G4_v11' of github.com:QTNM/QTNMSim into tom/build_G4_v11

yramachers commented:
@TomGoffrey I checked out the branch and it compiles nicely, without a hitch or warning. Starting without a macro also works -> expected warning and exit. The tests are commented out/empty. Maybe that could be another step before merging this pull request? Unless there is still a feature you wanted to include before merging?

yramachers committed:
[code] send gdml file name as std::string

yramachers committed:
[primary] add a test electron; 90 degree emission to default B-field from origin.

yramachers committed:
[userlimit] added user limit on maximum time for transport, attached to world volume.

yramachers committed:
[geometry] typo removed, need a geometry naming convention, names are in source code currently.

yramachers commented:
Compiles fine but a single electron test run produces a segmentation fault. Geometry appears to be read fine. Back to debugging.

yramachers committed:
[debug] G4cout debugging, compiles and seg faults at the end; wrong transport in B-field - check units.

yramachers commented:
The only connection I could find between dynamic particle and its properties and the eqn of motion 'set charge and mass' method is in the trajectory -> set charge and mass (and momentum, not used) in constructor using pointer to EqnOfMotion. Seems to work, but now the transport in the B-field is wrong. Main suspect is that units are wrong for charge and mass, currently using G4 internal units, likely wrong for that purpose. Checking is possible with plenty of G4cout statements still included. The segmentation fault is connected to the final output to ROOT. Lower priority, to be fixed later.

yramachers committed:
[debug] stopped trajectory outside volumes of interest and Boris stepper strange behaviour.

yramachers committed:
[debug] now use G411.1 BorisDriver, follow Field01 example, needs transport limits, check units.

yramachers committed:
[debug] working on output seg fault, not finished, cleanup first, unit check tbd.

yramachers committed:
[event] filling ntuple on hit causes seg fault. Continue debugging.

yramachers committed:
[output] new central Output Manager class, compiles, not run yet.

yramachers committed:
[restructure] only output manages storage vectors, clean event/runAction. Compiles, not run yet.

yramachers committed:
[style] adjust style to AnaEx01 on destructor.

yramachers committed:
[typos] edit typos and check includes. Compiles fine.

yramachers committed:
[analysis] first working version without seg fault, output to root works. Units next.

yramachers committed:
[eqn] adapt formulae to basics for unit checks. Field calculations next, not right yet.

yramachers committed:
[units] explicit SI units mostly on constants, checking formulae, sensible results.

yramachers committed:
[units] prefactor correction done. Check output values for test case.

yramachers committed:
[readme] updated readme and more prints on debug numbers.

yramachers committed:
[trajectory] better dipole antenna orientation, check voltage response.

yramachers committed:
[trajectory] normalised antenna polarisation vector.

yramachers committed:
[cleanup] removed debug prints, cleaned static constants, ready to merge.

yramachers committed:
[cleanup] no bespoke Boris solver at this stage, use G4Boris solvers.

yramachers reviewed (approved):
good point to merge and implement features separately later, each by separate branches.
